packages/NodeTypeResolver/PHPStan/Scope/PHPStanNodeScopeResolver.php

```diff
@@ -119,12 +119,40 @@ private function processNodesWithDependentFiles(
         MutatingScope $mutatingScope,
         callable $nodeCallback
     ): array {
+        $countDeepStmts = $this->countDeepStmts($stmts);
+
+        if ($countDeepStmts >= 20) {
+            return $stmts;
+        }
+
         $this->nodeScopeResolver->processNodes($stmts, $mutatingScope, $nodeCallback);
         $this->resolveAndSaveDependentFiles($stmts, $mutatingScope, $smartFileInfo);
 
         return $stmts;
     }
 
+    /**
+     * @param Stmt[] $stmts
+     */
+    private function countDeepStmts(array $stmts, int $lastCount = 0): int
+    {
+        foreach ($stmts as $stmt) {
+            if (! property_exists($stmt, 'stmts')) {
+                return $lastCount;
+            }
+
+            ++$lastCount;
+            if (! isset($stmt->stmts[0])) {
+                return $lastCount;
+            }
+
+            $stmt = $stmt->stmts[0];
+            $lastCount = $this->countDeepStmts([$stmt], $lastCount);
```

Comment:
@TomasVotruba this recursive count is assume that the stmts is 1 statement per-level, multi statement per-level will make count invalid.

The issue is actually seems on PHPStan :

```php
 $this->nodeScopeResolver->processNodes($stmts, $mutatingScope, $nodeCallback);
```